server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java

```diff
@@ -200,6 +177,43 @@ protected Response processExtensions(SMTPSession session, Mail mail) {
         return null;
     }
 
+    private Optional<SMTPResponse> getRawHandlersSmtpResponse(SMTPSession session, Mail mail, OutputStream out, MessageHook rawHandler) {
+        LOGGER.debug("executing james message handler {}", rawHandler);
+        long start = System.currentTimeMillis();
+
+        HookResult hRes = rawHandler.onMessage(session, new MailToMailEnvelopeWrapper(mail, out));
+        long executionTime = System.currentTimeMillis() - start;
+
+        if (rHooks != null) {
+            for (HookResultHook rHook : rHooks) {
+                LOGGER.debug("executing hook {}", rHook);
+                hRes = rHook.onHookResult(session, hRes, executionTime, rawHandler);
```

Comment:
Hook execution can be long (and blocking), we might need to wrap the entire method `getRawHandlersSmtpResponse` in a mono too be able to specify a rate here?